### ISSUE TITLE
Akka Cluster can get stuck joining re-started node when

### ISSUE THREAD

**Issue by ktoso** [akka/akka#19274]
I debugged this from some logs we gathered with @patriknw.

We believe this is some extremely rare condition in which (maybe, we're not sure yet) a node which is removed from the cluster and restarts, begins JOINING, gets accepted, the `InitJoinAck` is sent but lands in `deadLetters` and then the node wants to join again... and the cluster lets it, wanting to replace the old node with the same address with this one and somehow gets stuck in this state:

The problem looks like this (IPs and details redacted):
```
2015-12-20 03:40:27,403 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-14, akkaSource=Cluster(akka://ExampleService), akkaTimestamp=03:40:27.403UTC] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - Node [akka.tcp://ExampleService@3.3.3.3:2552] is JOINING, roles [a, b, c]
2015-12-20 03:41:21,981 WARN  [a.remote.ReliableDeliverySupervisor] [sourceThread=ExampleService-akka.remote.default-remote-dispatcher-5, akkaTimestamp=03:41:21.978UTC, akkaSource=akka.tcp://ExampleService@2.2.2.2:2552/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FExampleService%403.3.3.3%3A2552-16] - Association with remote system [akka.tcp://ExampleService@3.3.3.3:2552] has failed, address is now gated for [5000] ms. Reason: [Disassociated]
2015-12-20 03:41:21,982 WARN  [a.remote.ReliableDeliverySupervisor] [sourceThread=ExampleService-akka.remote.default-remote-dispatcher-5, akkaTimestamp=03:41:21.982UTC, akkaSource=akka.tcp://ExampleService@2.2.2.2:2552/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FExampleService%403.3.3.3%3A2552-17] - Association with remote system [akka.tcp://ExampleService@3.3.3.3:2552] has failed, address is now gated for [5000] ms. Reason: [Disassociated]
2015-12-20 03:41:27,831 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-18, akkaSource=Cluster(akka://ExampleService), akkaTimestamp=03:41:27.831UTC] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - New incarnation of existing member [Member(address = akka.tcp://ExampleService@3.3.3.3:2552, status = Joining)] is trying to join. Existing will be removed from the cluster and then new member will be allowed to join.
2015-12-20 03:41:27,832 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-18, akkaTimestamp=03:41:27.832UTC, akkaSource=Cluster(akka://ExampleService)] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - Marking node [akka.tcp://ExampleService@3.3.3.3:2552] as [Down]
2015-12-20 03:41:38,125 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-18, akkaTimestamp=03:41:38.125UTC, akkaSource=Cluster(akka://ExampleService)] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - New incarnation of existing member [Member(address = akka.tcp://ExampleService@3.3.3.3:2552, status = Down)] is trying to join. Existing will be removed from the cluster and then new member will be allowed to join.
2015-12-20 03:41:49,120 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-16, akkaSource=Cluster(akka://ExampleService), akkaTimestamp=03:41:49.120UTC] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - New incarnation of existing member [Member(address = akka.tcp://ExampleService@3.3.3.3:2552, status = Down)] is trying to join. Existing will be removed from the cluster and then new member will be allowed to join.
2015-12-20 03:42:00,110 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-15, akkaSource=Cluster(akka://ExampleService), akkaTimestamp=03:42:00.110UTC] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - New incarnation of existing member [Member(address = akka.tcp://ExampleService@3.3.3.3:2552, status = Down)] is trying to join. Existing will be removed from the cluster and then new member will be allowed to join.
2015-12-20 03:42:10,121 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-16, akkaSource=Cluster(akka://ExampleService), akkaTimestamp=03:42:10.121UTC] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - New incarnation of existing member [Member(address = akka.tcp://ExampleService@3.3.3.3:2552, status = Down)] is trying to join. Existing will be removed from the cluster and then new member will be allowed to join.
2015-12-20 03:42:21,113 INFO  [Cluster(akka://ExampleService)] [sourceThread=ExampleService-dispatchers.cluster-14, akkaTimestamp=03:42:21.113UTC, akkaSource=Cluster(akka://ExampleService)] - Cluster Node [akka.tcp://ExampleService@2.2.2.2:2552] - New incarnation of existing member [Member(address = akka.tcp://ExampleService@3.3.3.3:2552, status = Down)] is trying to join. Existing will be removed from the cluster and then new member will be allowed to join.
```

Was not yet able to reproduce this problem.
This was on Akka 2.3.

**Comment by ktoso** [comment]
(This was Akka 2.3)

**Comment by ktoso** [comment]
To be included in next 2.3 release, and 2.4.2.
